### PR TITLE
Feature 64 support auth 2

### PR DESCRIPTION
**Pull request by rushirajnenuji**
Requirements defined as part of #64:
> The R client needs to be able to:
>
>    store OAuth2 access tokens
>    store OAuth2 refresh tokens
>    send API calls to the vegbank API including Bearer tokens in the header
>    call keycloak's OpenID api to refresh access tokens when they expire (or are about to)
>    provide information to the user on how to login to get access and refresh tokens
>    notify the user when login is needed again

Builds on top of ([feature-support-auth-MVP](https://github.com/NCEAS/vegbankr/tree/feature-64-support-auth-mvp) -- related [draft PR](https://github.com/NCEAS/vegbankr/pull/76)) which can be closed without merging, since all changes are already included here and target is develop. 

Additional changes added via this PR:
- Adds support for `POST` to `refresh` endpoint
- Adds helpers for setting / clearing access and refresh tokens
    - accepts individual token strings as well as a JSON object (response from `vegbankapi`)
- Adds support for JWT claim extraction and token validation
    - checks token validation before sending requests
    - if a valid refresh token is available, obtains a new access and refresh token set; and proceeds with the request. Send a message to the user otherwise 
- All API requests are currently authenticated if a valid token is present
    - TODO: just modify this to support POST with a feature flag?
- Updates documentation with related examples
- Adds minimal set of tests for related methods

Closes out issues #64 and #66

### CODE REVIEW

File: R/vb_auth.R [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+#' the token exists and has a valid expiry.
+#'
+#' @param token JWT string, or NULL
+#' @returns Logical `TRUE` if the token is present and valid,
+#'   `FALSE` otherwise.
+#' @noRd
+token_is_valid <- function(token) {
+  exp <- jwt_exp(token)
+
+  # Keep a 30 second buffer for request round trip
```
Comment (rushirajnenuji) [inline R/vb_auth.R]: I can use some thoughts on this, I believe if we run into this edge case the `vegbankapi` should automatically try to update token if refresh token is available, but wasn't sure and added this just to be safe.

File: R/utils.R [hunk 2]
```diff
@@ -128,6 +130,26 @@ send <- function(request) {
...
+        vb_refresh_tokens()
+      } else {
+        message("Access token expired and no valid refresh token available.")
+        stop("Re-authenticate and call vb_set_token() to set a new token.",
+             call. = FALSE)
+      }
+    }
+
+    token <- vb_token()
+    if (!is.null(token) && vb_access_token_is_valid()) {
```
Comment (mbjones) [inline R/utils.R:148]: There is a logical case missing here, where if the `vb_refresh_tokens()` fails to return a valid token, for example if the refresh token itself is expired or revoked. In those cases, we should abort the send and raise an error because it can't be completed with proper credentials. If the user wants to retry the operation as the public user, they should unset their options of use `skip_auth=true`. So, we probably need an `else` block here, but only for the `vb_access_token_is_valid()` part of the condition.

File: R/vb_auth.R [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+  if (!is.null(tokens)) {
+    tokens        <- parse_tokens_dict(tokens)
+    access_token  <- tokens[["access_token"]]
+    refresh_token <- tokens[["refresh_token"]]
+  }
+
+  assert_token_string(access_token,  "access_token")
+  assert_token_string(refresh_token, "refresh_token")
+
+  if (is.null(access_token) && is.null(refresh_token)) {
```
Comment (mbjones) [inline R/vb_auth.R:38]: Shouldn't this be an `||` OR test? Seems odd that both would need to be null to fail the test.

File: R/vb_auth.R [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+#' * **Token dict:** pass `tokens` as a named list or JSON string with `access_token` and/or `refresh_token` keys.
+#'
+#' @param access_token (character) Access token string.
+#' @param refresh_token (character) Refresh token string.
+#' @param tokens (list | character) Named list or JSON string containing
+#'   `access_token` and/or `refresh_token`. Cannot be combined with
+#'   `access_token` or `refresh_token`.
+#' @returns NULL
+#' @examples
+#' vb_set_token(access_token = "eyJhbGciOiJIUzI1NiJ9...")
```
Comment (mbjones) [inline R/vb_auth.R:16]: Is it logically possible to set only an access token and not a refresh token? I saw logic elsewhere that I thought required both. But it does make sense to me that you always need at least an access token, and that a refresh token could be optional. If a refresh token is omitted, we probably should emit a message indicating that their token will expire in xx minutes and that providing a refresh token will allow it to be automatically renewed.

File: R/vb_auth.R [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+    tokens        <- parse_tokens_dict(tokens)
+    access_token  <- tokens[["access_token"]]
+    refresh_token <- tokens[["refresh_token"]]
+  }
+
+  assert_token_string(access_token,  "access_token")
+  assert_token_string(refresh_token, "refresh_token")
+
+  if (is.null(access_token) && is.null(refresh_token)) {
+    stop("at least one of 'access_token' or 'refresh_token' must be provided")
```
Comment (mbjones) [inline R/vb_auth.R:39]: Is this message correct? If we only provide a refresh token, is that sufficient?

File: R/vb_auth.R [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+  options(vegbank.refresh_token = NULL)
+  message("VegBank token(s) cleared")
+}
+
+
+#' Retrieve the currently stored Bearer token
+#'
+#' @returns The token string, or NULL if none is set
+#' @noRd
+vb_token <- function() {
```
Comment (mbjones) [inline R/vb_auth.R]: Maybe rename this to `vb_access_token` to be explicit it is only that?

File: R/vb_auth.R [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+#' [vb_set_token()]. After calling this function, subsequent API requests
+#' will be sent without an `Authorization: Bearer` header.
+#'
+#' @returns NULL
+#' @examples
+#' vb_unset_token()
+#' @seealso [vb_set_token()]
+#' @export
+vb_unset_token <- function() {
+  options(vegbank.token         = NULL)
```
Comment (mbjones) [inline R/vb_auth.R]: Should this be named `vegbank.access_token` to be parallel and explicit?

File: R/vb_auth.R [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+
+  # Decode payload to JSON string, then extract the requested claim
+  payload_json <- tryCatch(
+    rawToChar(base64enc::base64decode(payload_b64)),
+    error = function(e) NULL
+  )
+
+  if (is.null(payload_json)) return(NULL)
+  tryCatch(jsonlite::fromJSON(payload_json)[[claim]], error = function(e) NULL)
+}
```
Comment (mbjones) [inline R/vb_auth.R:138]: This is fine, but consider whether it is better to use the JOSE package (https://r-lib.r-universe.dev/jose) which includes decoding routines, token validation, and claim manipulation.

File: R/vb_auth.R [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+
+#' Extract the expiry time from a JWT string
+#'
+#' Convenience wrapper around [jwt_claim()] that extracts the `exp` claim
+#' and converts it to a POSIXct timestamp.
+#'
+#' @param token JWT string
+#' @returns POSIXct expiry time, or NULL
+#' @noRd
+jwt_exp <- function(token) {
```
Comment (mbjones) [inline R/vb_auth.R]: Same, consider JOSE

File: R/vb_auth.R [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+#' Check whether a token is present and not expired
+#'
+#' Decodes the `exp` claim from a JWT token string and returns `TRUE` if
+#' the token exists and has a valid expiry.
+#'
+#' @param token JWT string, or NULL
+#' @returns Logical `TRUE` if the token is present and valid,
+#'   `FALSE` otherwise.
+#' @noRd
+token_is_valid <- function(token) {
```
Comment (mbjones) [inline R/vb_auth.R:165]: Maybe use JOSE for validation? Plus, there is more to validation than just expiration time (also requires some valid claims and a good signature). Maybe consider naming this `token_is_expired`?

File: R/vb_auth.R (hunk 9 above)
Comment (mbjones) [reply R/vb_auth.R]: Maybe use a fuller name, `jwt_exp` is cryptic. Maybe `jwt_expires` or `jwt_exp_time`?

File: R/vb_auth.R (hunk 1 above)
Comment (mbjones) [reply R/vb_auth.R]: why is the buffer in the `is_valid` function? Seems like this function should return TRUE only if it is not expired at all. Any buffering should be based on the function that is doing the refresh renewal.

File: README.md [hunk 11]
```diff
@@ -116,6 +116,98 @@ sequoia_plots <- sequoia_communities |>
...
+    start_date = "2026-03-01"
+  ), dry_run = TRUE)
+```
+
+### Token lifecycle
+
+**Automatic refresh.** When an API request is made and the access token
+has expired, `vegbankr` will automatically use the stored refresh token
+to obtain a new token pair — no manual intervention required. If the
+refresh token has also expired, you will be prompted to re-authenticate.
```
Comment (mbjones) [inline README.md:187]: Is this true? I didn't catch the mechanism for the reprompt to login, but its a good feature.

File: R/utils.R [hunk 12]
```diff
@@ -128,6 +130,26 @@ send <- function(request) {
...
 
+  if (!skip_auth) {
+    # If the access token is expired but the refresh token is
+    # still valid, update tokens before retrying the request.
+    if (!is.null(vb_token()) && !vb_access_token_is_valid()) {
+      if (vb_refresh_token_is_valid()) {
+        message("Access token expired; refreshing tokens...")
+        vb_refresh_tokens()
+      } else {
+        message("Access token expired and no valid refresh token available.")
```
Comment (mbjones) [inline R/utils.R]: Provide the URL that they need to login in the message.

File: R/utils.R [hunk 13]
```diff
@@ -128,6 +130,26 @@ send <- function(request) {
   }
   request <- request |> req_error(body = error_body)
 
+  if (!skip_auth) {
+    # If the access token is expired but the refresh token is
+    # still valid, update tokens before retrying the request.
+    if (!is.null(vb_token()) && !vb_access_token_is_valid()) {
+      if (vb_refresh_token_is_valid()) {
+        message("Access token expired; refreshing tokens...")
+        vb_refresh_tokens()
```
Comment (mbjones) [inline R/utils.R:139]: Is this the only place `vb_refresh_tokens()` is called?

File: R/utils.R [hunk 14]
```diff
@@ -128,6 +130,26 @@ send <- function(request) {
   }
   request <- request |> req_error(body = error_body)
 
+  if (!skip_auth) {
+    # If the access token is expired but the refresh token is
+    # still valid, update tokens before retrying the request.
+    if (!is.null(vb_token()) && !vb_access_token_is_valid()) {
```
Comment (mbjones) [inline R/utils.R]: Should we add an expiration time check here? e.g., 

```
if (!is.null(vb_token()) && (!vb_access_token_is_valid() || (Sys.time() + 30) < jwt_exp() )) {
```

**Review by mbjones — APPROVED** [review]
@rushirajnenuji nice looking work. I'm getting my comments to you now but have not yet tried running it. I will pull and run shortly and provide more comments if anything arises.

**Comment by mbjones** [comment]
One thing I just noticed: the methods `vb_refresh_token()` and `vb_refresh_tokens()` are too closely named. We need to differentiate them.

How about the methods be called: `vb_get_token()`, `vb_get_refresh_token()`, and `vb_refresh_tokens()`?

File: R/utils.R (hunk 13 above)
Comment (rushirajnenuji) [reply R/utils.R:139]: yep, this `utils:send()` method is the only place where we use `vb_refresh_tokens()` -- gets applied to all requests currently.

File: R/utils.R (hunk 12 above)
Comment (rushirajnenuji) [reply R/utils.R]: resolved with https://github.com/NCEAS/vegbankr/pull/77/changes/6a97031f12102046760e6d3fafbbee23dc3f084a

File: R/utils.R (hunk 14 above)
Comment (rushirajnenuji) [reply R/utils.R]: I like the idea of moving this to `send()`. Updated with https://github.com/NCEAS/vegbankr/pull/77/changes/9e2d3f671ad2d292918e5ae6c99157ad63e8f145

File: R/vb_auth.R (hunk 9 above)
Comment (rushirajnenuji) [reply R/vb_auth.R]: resolved; renamed to `jwt_expiry_time()` with https://github.com/NCEAS/vegbankr/pull/77/changes/9e2d3f671ad2d292918e5ae6c99157ad63e8f145

File: R/vb_auth.R (hunk 7 above)
Comment (rushirajnenuji) [reply R/vb_auth.R]: resolved with https://github.com/NCEAS/vegbankr/pull/77/changes/9e2d3f671ad2d292918e5ae6c99157ad63e8f145

File: R/vb_auth.R (hunk 6 above)
Comment (rushirajnenuji) [reply R/vb_auth.R]: resolved with https://github.com/NCEAS/vegbankr/pull/77/changes/9e2d3f671ad2d292918e5ae6c99157ad63e8f145

File: R/vb_auth.R (hunk 1 above)
Comment (rushirajnenuji) [reply R/vb_auth.R]: buffer is now moved to `send()` method as recommended. https://github.com/NCEAS/vegbankr/pull/77/changes/9e2d3f671ad2d292918e5ae6c99157ad63e8f145

File: R/vb_auth.R (hunk 5 above)
Comment (rushirajnenuji) [reply R/vb_auth.R:39]: sort of, the idea is that if a refresh token is provided, the logic would detect that access token is missing, and use the refresh token to get updated tokens. But I see the point that we might never have just a refresh token.

File: R/utils.R (hunk 12 above)
Comment (rushirajnenuji) [reply R/utils.R]: resolved with https://github.com/NCEAS/vegbankr/pull/77/changes/9e2d3f671ad2d292918e5ae6c99157ad63e8f145

File: R/vb_auth.R (hunk 4 above)
Comment (rushirajnenuji) [reply R/vb_auth.R:16]: resolved with https://github.com/NCEAS/vegbankr/pull/77/changes/9e2d3f671ad2d292918e5ae6c99157ad63e8f145
checks expiry and returns a message with expiry time if only access_token is provided.

**Comment by rushirajnenuji** [comment]
Thank you @mbjones for thorough feedback and testing. 🙏🏻🙌🏻

I've tried to address some suggestions as part of this PR, and created additional tickets for the rest.

File: R/vb_auth.R (hunk 3 above)
Comment (mbjones) [reply R/vb_auth.R:38]: Looking at this more closely, I think the logic is fine (I had thought it was testing "not null", but with this checks if both tokens are null, which works.

File: R/vb_auth.R (hunk 5 above)
Comment (mbjones) [reply R/vb_auth.R:39]: lgtm

File: R/utils.R (hunk 2 above)
Comment (mbjones) [reply R/utils.R:148]: resolved in issue #81 and PR #104
